internal/datanode/flow_graph_time_tick_node.go

```diff
@@ -72,43 +83,44 @@ func (ttn *ttNode) IsValidInMsg(in []Msg) bool {
 // Operate handles input messages, implementing flowgraph.Node
 func (ttn *ttNode) Operate(in []Msg) []Msg {
 	fgMsg := in[0].(*flowGraphMsg)
+	curTs, _ := tsoutil.ParseTS(fgMsg.timeRange.timestampMax)
 	if fgMsg.IsCloseMsg() {
 		if len(fgMsg.endPositions) > 0 {
+			ttn.closeChannel <- true
 			channelPos := ttn.channel.getChannelCheckpoint(fgMsg.endPositions[0])
 			log.Info("flowgraph is closing, force update channel CP",
 				zap.Time("cpTs", tsoutil.PhysicalTime(channelPos.GetTimestamp())),
 				zap.String("channel", channelPos.GetChannelName()))
-			ttn.updateChannelCP(channelPos)
+			ttn.updateChannelCP(channelPos, curTs)
 		}
 		return in
 	}
 
-	curTs, _ := tsoutil.ParseTS(fgMsg.timeRange.timestampMax)
+	// Do not block and async updateCheckPoint
 	channelPos := ttn.channel.getChannelCheckpoint(fgMsg.endPositions[0])
-	log := log.With(zap.String("channel", ttn.vChannelName),
-		zap.Time("cpTs", tsoutil.PhysicalTime(channelPos.GetTimestamp())))
-	if curTs.Sub(ttn.lastUpdateTime) >= updateChanCPInterval {
-		if err := ttn.updateChannelCP(channelPos); err == nil {
-			ttn.lastUpdateTime = curTs
-			log.Info("update channel cp periodically")
-			return []Msg{}
+	if curTs.Sub(ttn.lastUpdateTime.Load()) >= updateChanCPInterval {
+		select {
+		case ttn.notifyChannel <- checkPoint{curTs, channelPos}:
+		default:
 		}
```

Comment:
Please don't repeat yourself, make a func of it.

Reply:
updated